components/core/src/reducer/ServerContext.cpp

```diff
@@ -0,0 +1,255 @@
+        : m_tcp_acceptor{m_ioctx, tcp::endpoint(tcp::v4(), args.get_reducer_port())},
+          m_scheduler_socket{m_ioctx},
+          m_upsert_timer{m_ioctx},
+          m_reducer_host{args.get_reducer_host()},
+          m_reducer_port{args.get_reducer_port()},
+          m_polling_interval_ms{args.get_polling_interval()} {
+    mongocxx::uri mongodb_uri = mongocxx::uri(args.get_mongodb_uri());
+    try {
+        m_mongodb_client = mongocxx::client(mongodb_uri);
+    } catch (mongocxx::exception& e) {
+        SPDLOG_ERROR("Failed to connect to {} - {}", args.get_mongodb_uri(), e.what());
+        throw OperationFailed(clp::ErrorCode_BadParam, __FILENAME__, __LINE__);
+    }
+    m_mongodb_results_database = mongocxx::database(m_mongodb_client[mongodb_uri.database()]);
+}
+
+void ServerContext::reset() {
+    m_ioctx.restart();
+    m_pipeline.reset(nullptr);
+    m_status = ServerStatus::Idle;
+    m_job_id = -1;
+    m_is_timeline_aggregation = false;
+    m_updated_tags.clear();
+    m_num_active_receiver_tasks = 0;
+}
+
+void ServerContext::stop_event_loop() {
+    m_tcp_acceptor.cancel();
+    m_scheduler_socket.close();
+}
+
+bool ServerContext::register_with_scheduler(
+        boost::asio::ip::tcp::resolver::results_type const& endpoints
+) {
+    try {
+        boost::asio::connect(m_scheduler_socket, endpoints);
+    } catch (boost::system::system_error& error) {
+        SPDLOG_ERROR("Failed to connect to search scheduler - {}", error.what());
+        return false;
+    }
+
+    boost::system::error_code error;
+
+    nlohmann::json reducer_advertisement;
+    reducer_advertisement["host"] = m_reducer_host;
+    reducer_advertisement["port"] = m_reducer_port;
+    auto serialized_advertisement = nlohmann::json::to_msgpack(reducer_advertisement);
+    auto message_size = serialized_advertisement.size();
+
+    boost::asio::write(
+            m_scheduler_socket,
+            boost::asio::buffer(&message_size, sizeof(message_size)),
+            error
+    );
+    if (error) {
+        m_scheduler_socket.close();
+        return false;
+    }
+
+    boost::asio::write(m_scheduler_socket, boost::asio::buffer(serialized_advertisement), error);
+    if (error) {
+        m_scheduler_socket.close();
+        return false;
+    }
+
+    return true;
+}
+
+bool ServerContext::ack_search_scheduler() {
+    boost::system::error_code e;
+    char const scheduler_response = 'y';
+    boost::asio::write(m_scheduler_socket, boost::asio::buffer(&scheduler_response, 1), e);
+    if (e) {
+        SPDLOG_ERROR("Failed to notify search scheduler - {}", e.message());
+        return false;
+    }
+    return true;
+}
+
+void ServerContext::decrement_num_active_receiver_tasks() {
+    --m_num_active_receiver_tasks;
+    if (0 == m_num_active_receiver_tasks && ServerStatus::ReceivedAllResults == m_status) {
+        if (false == try_finalize_results()) {
+            m_status = ServerStatus::UnrecoverableFailure;
```

Comment:
Are all the callers of this method handling the `UnrecoverableFailure`? Perhaps we should move the status setting out of the method for clarity (the same argument you had for keeping the status updates outside of the class)?

Reply:
Not directly, but the intention is that every routine should exit as soon as this happens, and then we log and exit in main. The flow is a bit convoluted, but this behaviour is guaranteed because every task either checks status and exits on unexpected status or gets cancelled/terminates when ServerStatus::ReceivedAllResults is set.